[WIP] - Switch to Rollup

Florian-R opened the pull request:
This is one should fixes #894. For now, all tests explode because I couldn't figure a way to compile the code correctly with grunt-mocha-test, but I'd like to grab early comments.

TBD:

- [x] Fix tests
- [x] Tests on real world app to see it fixes Brunch and Gulp issues
- [x] Check for deps require
- [x] Check bundle size - Generated with grunt-compare-size : 
```
-255216 build/chaplin.js
-4992 build/chaplin.min.js
```
- [x] If possible, get rid of coffee extensions (see inline comments)
- [x] Check if coverage is still OK.

cc @embs @shvaikalesh

Florian-R commented on src/chaplin/composer.coffee:
```diff
@@ -3,9 +3,9 @@
 _ = require 'underscore'
 Backbone = require 'backbone'
```
I've just stumble upon the `external` option of rollup, I'll change this one in a future commit

Florian-R commented on Gruntfile.coffee:
```diff
@@ -61,8 +37,9 @@ module.exports = (grunt) ->
+            # 'coffee-coverage/register-istanbul'
             'coffee-script/register'
-            'coffee-coverage/register-istanbul'
+            'babel-register'
```
Tests breaks, because we babel-register had a [extension hook](https://nodejs.org/api/globals.html#globals_require_extensions) but it expect `.js` file.

A simpler approach could be compile our files before running test in a tmp dir, but it may breaks coverage and will be slower and dirtier.

Florian-R commented on src/chaplin.coffee:
```diff
@@ -1,23 +1,43 @@
 'use strict'
 
+import Application from './chaplin/application.coffee'
```
Had to keep extensions, else rollup doesn't resolve the files. See rollup/rollup#1053

shvaikalesh commented on src/chaplin.coffee:
```diff
@@ -1,23 +1,43 @@
+  CollectionView
+  Layout
+  View
+}
```
ES6 modules can do `export { foo } from 'bar.coffee'`, but that won't help here. Furthermore, we should be careful not break CommonJS users because of `default` thing.

paulmillr commented:
👍

Florian-R commented:
Added some commits, I got rid of the coffee extensions, and cleaned remaining `require`.

I've updated my first post, the bundle is way smaller (didn't check gzip tough).

I'd work on getting the tests to pass, in the meantime, @embs could you do a quick check with my [rollup-build branch](https://github.com/Florian-R/chaplin/tree/rollup-build) and confirm it fixes your gulp setup? I can confirm this one works fine with Brunch.

shvaikalesh commented:
Outstanding job, @Florian-R! Code looks so much better now.

shvaikalesh commented on src/chaplin/application.coffee:
```diff
@@ -106,7 +106,7 @@ module.exports = class Application
   # is a replacement for Backbone.Router and *does not inherit from it*
   # directly. It's a different implementation with several advantages over
   # the standard router provided by Backbone. The router is typically
-  # initialized by passing the function returned by **routes.coffee**.
```
Not sure that we should remove that.

shvaikalesh commented:
Also, we should remove all `'use strict'` declarations, because es6 modules are always strict.

vendethiel commented on src/chaplin/lib/utils.coffee:
```diff
@@ -1,5 +1,7 @@
 'use strict'
 
+import mediator from '../mediator'
```
No, the require was intended. There's a dependency cycle between utils and mediator.

Florian-R commented:
> Also, we should remove all 'use strict' declarations, because es6 modules are always strict.

Make sense, done for the `src` directory for now.

Florian-R commented:
Added one commits for the test. I had to use a tmp dir to compile the coffee. See benbria/coffee-coverage/issues/82 for the rationale. Let me know if your good with this approach, and I'll tweak the code to avoid a double compilation.

Florian-R commented on Gruntfile.coffee:
```diff
@@ -298,17 +286,25 @@ module.exports = (grunt) ->
-  grunt.registerTask 'test', 'mochaTest:native'
-  grunt.registerTask 'test:jquery', 'mochaTest:jquery'
+  grunt.registerTask 'test', ['clean', 'instrument', 'coffee:test', 'mochaTest:native']
+  grunt.registerTask 'test:jquery', ['clean', 'instrument', 'coffee:test', 'mochaTest:jquery']
```
This part need some rework.

Florian-R commented on package.json:
```diff
@@ -37,5 +42,10 @@
+    "plugins": [
+      "transform-es2015-modules-commonjs"
+    ]
+  }
```
Inlined this to avoid a .babelrc

Florian-R commented on Gruntfile.coffee:
```diff
@@ -56,51 +32,63 @@ module.exports = (grunt) ->
-            'coffee-script/register'
-            'coffee-coverage/register-istanbul'
+            ->
+              global._$coffeeIstanbul = {}
```
This is ugly, but needed. `istanbul` add something like 

```js
if (typeof _$coffeeIstanbul === 'undefined') _$coffeeIstanbul = {};
```

in every modules. This breaks badly with `babel-register` which wrap the code in a function with a `use strict` pragma.

Florian-R commented on package.json:
```diff
@@ -4,30 +4,35 @@
     "grunt-cli": "~0.1.13",
     "grunt-coffeelint": "~0.0.15",
+    "grunt-contrib-clean": "1.0.0",
+    "grunt-contrib-coffee": "florian-r/grunt-contrib-coffee",
```
This is WIP pending a new release of `grunt-contrib-coffee`. The current version use an old version of `coffee-script`.

Florian-R commented:
Closing per #900
